src/viewmodels/base/useAutoDisposedViewModel.ts

```diff
@@ -0,0 +1,62 @@
+/*
+Copyright 2025 New Vector Ltd.
+SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only OR LicenseRef-Element-Commercial
+Please see LICENSE files in the repository root for full details.
+*/
+
+import { useEffect, useState } from "react";
+
+import type { BaseViewModel } from "./BaseViewModel";
+
+type VmCreator<B extends BaseViewModel<unknown, unknown>> = () => B;
+
+/**
+ * Instantiate a view-model that gets disposed when the calling react component unmounts.
+ * In other words, this hook ties the lifecycle of a view-model to the lifecycle of a
+ * react component.
+ *
+ * @param vmCreator A function that returns a view-model instance
+ * @returns view-model instance from vmCreator
+ * @example
+ * const vm = useAutoDisposedViewModel(() => new FooViewModel({prop1, prop2, ...});
+ */
+export function useAutoDisposedViewModel<B extends BaseViewModel<unknown, unknown>>(vmCreator: VmCreator<B>): B {
+    /**
+     * The view-model instance may be replaced by a different instance in some scenarios.
+     * We want to be sure that whatever react component called this hook gets re-rendered
+     * when this happens, hence the state.
+     */
+    const [viewModel, setViewModel] = useState<B>(vmCreator);
```

Comment:
If we're okay with the view-model returned from this hook being undefined, we could just delay creating the vm to the effect body below. But we would have to always `vm && <FooView vm={vm}>` and `vm?.doSomething()`.

 It would also mean that all views would require an additional render (effect body always runs after render so the vm is only created after one render) before they are mounted. Also for view-models that immediately start doing  async operations on instantiation, they would be delayed by one render cycle. But I doubt any of these would have actual performance implications.

I personally favor this approach because it feels equivalent to just calling the vm ctor.